tensorboard/data/server/reservoir.rs

```diff
@@ -0,0 +1,329 @@
+    /// `staged_items`.
+    committed_steps: Vec<Step>,
+    /// Items currently in the reservoir but not yet committed. Stored in step-sorted order, and
+    /// all steps in `staged_items` succeed all steps in `committed_steps`.
+    staged_items: Vec<(Step, T)>,
+    /// Total capacity of this reservoir. The combined physical capacities of `committed_steps` and
+    /// `staged_items` may exceed this, but their combined lengths will not. Behavior is undefined
+    /// if `capacity == 0`.
+    capacity: usize,
+    /// Reservoir control, to determine whether and whither a given new record should be included.
+    ctl: C,
+    /// Total number of records passed in the stream so far, regardless of whether they were ever
+    /// added to the reservoir. Usually called `N` in the literature.
+    seen: usize,
+}
+
+/// A `ReservoirControl` determines which records from a stream should be included into a
+/// reservoir, and which records they should evict. This is usually backed by a random number
+/// generator, but may be made deterministic for testing.
+pub trait ReservoirControl {
+    /// Upon seeing the latest record in a stream of `n` so far, rolls for the index into the
+    /// reservoir that should be evicted to make room for this record. If the result is greater
+    /// than or equal to the capacity of the reservoir, the record will be skipped; otherwise, it
+    /// will evict the record at the given index. The input `n` will be positive, and the result
+    /// must be in `0..n`.
+    ///
+    /// A useful implementation is to simply roll a random integer in `0..n`. Test code may wish to
+    /// provide an alternate, deterministic implementation.
+    fn destination(&mut self, n: usize) -> usize;
+}
+
+impl<R: Rng> ReservoirControl for R {
+    fn destination(&mut self, n: usize) -> usize {
+        Uniform::from(0..n).sample(self)
+    }
+}
+
+impl<T> StageReservoir<T, ChaCha20Rng> {
+    /// Creates a new reservoir with the specified capacity, using a fixed-seed random number
+    /// generator for reservoir control. All reservoirs created by this function will use the same
+    /// sequence of random numbers.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn new(capacity: usize) -> Self {
+        Self::with_control(capacity, ChaCha20Rng::seed_from_u64(0))
+    }
+}
+
+impl<T, C: ReservoirControl> StageReservoir<T, C> {
+    /// Offers a record to the reservoir. The reservoir will always include the latest record.
+    /// Other than the latest record, the records kept form a simple random sample of the stream
+    /// (or at least approximately so in the case of preemptions).
+    pub fn offer(&mut self, step: Step, v: T) {
+        self.seen += 1;
+        let dst = self.ctl.destination(self.seen);
+
+        // Didn't make the cut? Keep-last only.
+        if dst >= self.capacity {
+            if self.staged_items.pop().is_none() {
```

Comment:
Maybe it's overkill if this is the only place we use them, but FWIW, I think it'd be slightly easier for me to grasp the sampling logic if we abstracted away the bookkeeping inherent in the bipartite approach into helper functions that make the StageReservoir behave in some respects like one big Vec.

E.g. we could have

```rust
fn pop(&mut self) -> Option<T>
fn len(&self) -> usize
fn remove(&mut self, index: usize)  // Deviate from Vec by returning nothing here
```

Each of these helpers would simplify one of the three commented cases here.

Also, once the cases are compact enough, it might be nice to skip the early-returns and consolidate the three separate instances of `self.staged_items.push((step, v));` like this:

```rust
if dst >= self.capacity {
    self.pop();
} else if self.len() >= self.capacity {
    self.remove(dst);
}
self.staged_items.push((step, v));
```